google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -189,17 +190,21 @@ def __del__(self):
         """
         logger.debug("Entering deconstructor")
 
-        if self._current is not None:
-            logger.debug("Waiting for _current_instance_data to finish")
-            self._current.cancel()
-
-        if not self._client_session.closed:
-            logger.debug("Waiting for _client_session to close")
-            close_future = asyncio.run_coroutine_threadsafe(
-                self.__client_session.close(), loop=self._loop
-            )
-            close_future.result()
-
+        async def _deconstruct():
+            if not self._client_session.closed:
+                logger.debug("Waiting for _client_session to close")
+                await self._client_session.close()
+            if self._current is not None:
+                logger.debug("Waiting for _current to be cancelled")
+                await complete_task_cancellation(self._current)
+            if self._next is not None:
+                logger.debug("Waiting for _next to be cancelled")
+                await complete_task_cancellation(self._next)
+
+        deconstruct_future = asyncio.run_coroutine_threadsafe(
+            _deconstruct(), loop=self._loop
+        )
+        deconstruct_future.result()
```

Comment:
We might want to handle this in a different PR, but I think we may have inadvertently started using an anti-pattern here and are forcing all of the `icm` close synchronously. 

What might be better is if we switch to having an object that holds all the references to the individual connection managers and closes them out in async, all at once.

Reply:
I'll open another issue and PR for this then and just remove the references to __del__ in this PR

Reply:
So it turns out that the `__del__` function was what was causing the tests to hang, so I added these changes back in, but I agree that we should close the connection managers out in async rather than sync calls for each one.

Reply:
#77